[Nago] - ListView UI 구현, GridView 기능 구현, 커스텀 훅 (useFetch) 구현

Nago730 opened the pull request:
# ✅ 완료 작업 목록  

## 🔹 GridView  

### 📌 구현 내용  
- `mockdata.json`을 `fetch`로 불러와 동적으로 렌더링  
- `mockdata.json` 데이터를 `props`로 전달할 수 있도록 변환하는 `newsTransFormater.ts` 함수 추가  
- GridView의 언론사 24개 단위 슬라이싱 기능 추가  
  - 24개 단위로 페이지 분할 (최대 4페이지)  
  - `PrevPageBtn`, `NextPageBtn` 버튼 추가  
  - 첫 페이지와 마지막 페이지에서 버튼 숨김  
- 비동기 이슈 해결을 위해 `useFetch` 커스텀 훅 도입  
- 언론사 카드 호버 시 "구독하기" 버튼 표시  
  - 구현 과정에서 `@emotion/babel-plugin` 도입  

### 📂 파일 구조  
```plaintext
📁 GridView  
 ├── Grid.tsx  
 ├── NextPageBtn.tsx  
 ├── PrevPageBtn.tsx  
```

---

## 🔹 ListView  

### 📌 구현 내용  
- ListView UI 컴포넌트 구조 설계 및 구현  
- 뉴스 카드 구성 요소 분리  

### 📂 파일 구조  
```plaintext
📁 ListView  
 ├── 📁 NewsCard  
 │   ├── 📁 NewsInfo  
 │   │   ├── BrandMark.tsx  
 │   │   ├── EditDate.tsx  
 │   │   ├── SubscribeBtn.tsx  
 │   │   ├── UnSubscribeBtn.tsx  
 │   ├── 📁 NewsContent  
 │   │   ├── 📁 Main  
 │   │   │   ├── Thumbnail.tsx  
 │   │   │   ├── Title.tsx  
 │   │   ├── 📁 Sub  
 │   │   │   ├── Title.tsx  
 │   │   │   ├── Caption.tsx  
 │   ├── index.tsx  
 ├── 📁 Header  
 │   ├── 📁 ProgressBtn  
 │   │   ├── Count.tsx  
 │   │   ├── Label.tsx  
 │   ├── TextBtn.tsx  
```

---

# 🛠 설계 과정에서의 주요 고민과 해결  

## 🔹 GridView  

### 🔍 문제 상황  
- `fetch`를 통해 mock 데이터를 가져오는 과정에서 비동기 이슈 발생  
  - 데이터가 아직 로드되지 않았는데 화면이 렌더링되면서 오류 발생  

### ✅ 해결 방안  
- 로딩 상태(`useState`)를 추가하여 데이터가 완전히 로드될 때까지 빈 화면 유지  
- `useFetch` 커스텀 훅을 도입하여 비동기 데이터 로딩을 효율적으로 처리  
- 상세 내용: [비동기 데이터 처리 및 커스텀 훅 활용](https://github.com/Nago730/FE-newsstand/wiki/%EB%B9%84%EB%8F%99%EA%B8%B0-%EB%8D%B0%EC%9D%B4%ED%84%B0-%EC%B2%98%EB%A6%AC-%EB%B0%8F-%EC%BB%A4%EC%8A%A4%ED%85%80-%ED%9B%85-%ED%99%9C%EC%9A%A9)  

---

## 🔹 ListView  

### 🔍 컴포넌트 분리 기준 고민  
- 컴포넌트를 얼마나 작은 단위로 쪼갤 것인가?  
  - 버튼 내 아이콘과 텍스트를 개별 컴포넌트로 분리할 것인가?  
  - 적절한 모듈화 기준 설정  

### ✅ 해결 방안  
- 최소 단위까지 컴포넌트를 분리하여 재사용성을 높임  
- 디렉토리 구조를 활용하여 관련 컴포넌트 그룹화 (`index.tsx` 활용)  

---

## 🔹 TypeScript에서 `new Date()`와 Styled-Component 충돌 문제  

### 🔍 문제 상황  
- `new Date()` 사용 시 TypeScript와 Styled-Component 간의 충돌 발생  

### ✅ 해결 방안  
- 명시적으로 `Date` 객체를 변환하여 타입 오류 방지  
- 상세 내용: [TypeScript ‐ new Date() 사용 시 충돌 오류 해결 방법](https://github.com/Nago730/FE-newsstand/wiki/TypeScript-%E2%80%90-new-Date()-%EC%82%AC%EC%9A%A9-%EC%8B%9C-%EC%B6%A9%EB%8F%8C-%EC%98%A4%EB%A5%98-%ED%95%B4%EA%B2%B0-%EB%B0%A9%EB%B2%95)  

---

## 🔹 Hover UI 구현 방식 비교  

### 🔍 문제 상황  
- 특정 영역에 마우스를 올리면 버튼을 표시하고, 아닐 때는 이미지 유지  
- 구현 방식:  
  1. `useState`를 활용한 React 상태 기반 방식  
  2. CSS `:hover` 속성을 활용한 스타일링 방식  

### ✅ 해결 방안  
- 성능, 코드 가독성 등을 비교한 결과, CSS `:hover` 방식을 선택  
- Emotion의 Babel 플러그인을 활용하여 구현  
- 상세 내용: [Emotion CSS Hover vs React State (`useState`) 방식 비교](https://github.com/Nago730/FE-newsstand/wiki/Emotion-CSS-Hover-vs-React-State-(%60useState%60)-%EB%B0%A9%EC%8B%9D-%EB%B9%84%EA%B5%90)

Nago730 committed:
feat: 뷰 전환 기능 구현

- ViewType 타입 정의 및 export
- MediaView에서 상태 관리 (currentView)
- TabAndViewer 컴포넌트에 뷰 전환 기능 추가
- ViewerBtns 스타일링 (active 상태에 따른 아이콘 색상 변경)

Nago730 committed:
docs: PR 템플릿 생성

Nago730 committed:
refactor: ViewerBtns 컴포넌트 개선

- ViewBtn과 GridBtn을 하나의 공통 컴포넌트로 통합
- type prop을 통한 버튼 구분 로직 구현
- SVG 아이콘 컴포넌트화 (vite-plugin-svgr 적용)

Nago730 committed:
feat: ListView 컴포넌트 UI 구현 및 목데이터 연동

- ListView 하위 컴포넌트 구조 구현
- 목데이터를 활용한 UI 렌더링 완성
- PressNews, PressList 등 하위 컴포넌트 마크업

Nago730 committed:
docs PR template 위치 변경

Nago730 committed:
feat: 현재 날짜를 포맷하여 출력하는 기능 추가

- getFormattedDate 함수 구현
- 현재 날짜를 "YYYY. M. D. 요일" 형식으로 반환
- 요일을 한글로 표시

Nago730 committed:
feat: mock 데이터 추가

- 개발 및 테스트를 위한 mock 데이터 파일 추가
- root 디렉토리에 mock 데이터 위치

Nago730 committed:
chore: 컴포넌트 네이밍 컨벤션 통일

crongro commented on src/components/MainLayout/Header/TodaysDate/index.tsx:
```diff
@@ -0,0 +1,27 @@
+  return <Date>{getFormattedDate()}</Date>
+}
+
+function getFormattedDate(): string {
```
이런함수에 어떤 파라미터가 있으면 좋을지 생각해보세요. 범용적으로 동작하는 유틸이니까.
그리고 아래 weekdays와 같은 고정된 변수는 함수 밖에서 선언해도 되고요.

crongro commented on src/components/MainLayout/Media/ListView/Header/ProgressBtn/index.tsx:
```diff
@@ -0,0 +1,26 @@
+import Count from './Count'
+import { ProgressBtnProps } from '../index'
+
+function ProgressBtn({ text, currentIndex, totalCount }: ProgressBtnProps) {
```
ProgressBtnProps 를 한 파일에 표현하는 경우도 많은데요.
지금처럼 분리하는것도 나쁘지 않아보이네요.

crongro commented on src/components/MainLayout/Media/ListView/Header/index.tsx:
```diff
@@ -0,0 +1,37 @@
+  return (
+    <Container>
+      <ProgressBtn text="종합/경제" currentIndex={1} totalCount={81} />
+      <TextBtn text="방송/통신" />
+      <TextBtn text="IT" />
+      <TextBtn text="영자지" />
+      <TextBtn text="스포츠/연예" />
+      <TextBtn text="매거진/전문지" />
+      <TextBtn text="지역" />
```
반복문으로 처리해도 될 듯요.

crongro reviewed (approved):
컴포넌트가 작게 나눠져 있어서 좋군요.
타입은 투입추론이 자동으로 되는 경우도 있을테니 모든 타입을 선언하진 마세요.
그리고 타입을 분리할때는 복잡한 경우에는 그럴수 있지만 간단한 타입선언은 사용하는 파일과 함께 위치하는것도 좋고요.

Nago730 committed:
chore: 컴포넌트 네이밍 컨벤션 통일

Nago730 committed:
feat: 동적 GridView 구현 및 페이지네이션 추가

- 데이터 fetch를 위한 useFetch 커스텀 훅 생성
- 데이터 가공을 위한 newsTransFormater 추가
- GridView에 이전/다음 페이지 버튼 구현
- mock 데이터 기반 동적 렌더링 방식으로 GridView 리팩토링

Nago730 committed:
feat: 구독 관련 버튼 컴포넌트 공통화 및 Grid 호버 효과 추가

- 구독/구독취소 버튼을 Common 컴포넌트로 분리
- Grid 컴포넌트에 호버 시 구독 버튼 표시 기능 구현
- @emotion/babel-plugin 추가하여 스타일 컴포넌트 기능 확장

Nago730 committed:
docs: /.github/pull_request_template.md 추가

crongro commented on src/components/MainLayout/Media/GridView/index.tsx:
```diff
@@ -1,20 +1,55 @@
-  background-color: rgba(210, 218, 224, 1);
-  padding: 1px;
+  border-top: 1px solid rgba(210, 218, 224, 1);
+  border-left: 1px solid rgba(210, 218, 224, 1);
```
1px solid rgba(210, 218, 224, 1)

자주 보이는데 css변수를 활용해보세요.

crongro commented on src/hooks/useFetch.ts:
```diff
@@ -0,0 +1,20 @@
+import { useState, useEffect } from 'react'
+
+function useFetch(url: string) {
```
커스텀훅! 좋습니다.
URL받는것도 좋고요. 
이 모듈을 2군데정도에서 사용해보면서 다듬어보세요.

crongro reviewed (approved):
짧게 리뷰

crongro commented on src/components/MainLayout/Media/GridView/index.tsx:
```diff
@@ -0,0 +1,55 @@
+
+function GridView({ data }: GridViewProps) {
+  const [currentPage, setCurrentPage] = useState(1)
+  const totalPages = Math.ceil(data.length / 24)
+
+  const pages = Array.from({ length: totalPages }, (_, i) => {
+    const start = i * 24
+    const end = start + 24
+    return data
+      .slice(start, end)
+      .map(item => <Grid key={item.pid} logoLight={item.logoLight} name={item.name} />)
```
다 좋은데 24라는 숫자도 변수화 시켜서 뭔지 표현하면 더 좋죠.

crongro commented on src/components/MainLayout/Media/ListView/NewsCard/NewsContent/Sub/index.tsx:
```diff
@@ -0,0 +1,28 @@
+
+  return (
+    <Container>
+      <Title text={titleText}></Title>
+      <Title text={titleText}></Title>
+      <Title text={titleText}></Title>
+      <Title text={titleText}></Title>
+      <Title text={titleText}></Title>
+      <Title text={titleText}></Title>
```
테스트라고 하여도, map으로 표현하는것이 좋겠어요
